### PR TITLE
Dynamo+LTC: merging related code from staging branch to master

### PR DESCRIPTION
**Pull request by shunting314**
Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.

Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

### CODE REVIEW

File: torch/_lazy/__init__.py [hunk 1]
```diff
@@ -20,3 +20,27 @@ def wait_device_ops(devices=None):
...
+    Sync the list of lazy tensors so there IR get lowered for the activate backend
+    and the compiled computation graph get cached.
+    """
+    torch._C._lazy._sync_multi(tensors, devices)
+
+def get_tensor_id(tensor):
+    """Return a unique id of the lazy tensor maintained by LTC"""
+    return torch._C._lazy._get_tensor_id(tensor)
+
+def get_tensors_text(tensors):
```
Comment (wconstab) [inline torch/_lazy/__init__.py:35]: i wonder if we want to make a _lazy.debug. module for apis like this,
and if we want to rename the APIs

get_tensors_text -> debug.dump_ir(tensors)?

get_tensors_dot -> debug.render_ir_graph(tensors)?

it's easier to land it as-is, i am ok with possibly renaming/moving in a separate PR.

File: torch/_lazy/computation.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import torch._C._lazy
+import torch._C._lazy_ts_backend
+
+def get_tensors_ts_device_data_node(tensors):
```
Comment (wconstab) [inline torch/_lazy/computation.py:4]: I think anything that is ts_backend specific should be under the torch._lazy.ts_backend module.

File: torch/_lazy/computation.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    """Return tensor ids and eager tensors for DeviceData nodes in the
+       IR for the passed in lazy tensors."""
+    return torch._C._lazy_ts_backend._get_tensors_ts_device_data_node(tensors)
+
+def get_graph_hash(tensors):
+    """Return the graph hash for the passed in lazy tensors"""
+    return torch._C._lazy._get_graph_hash(tensors)
+
+def run_cached_graph(hash_str, graph_inputs):
```
Comment (wconstab) [inline torch/_lazy/computation.py:13]: maybe for now you can leave this 'run_cached_graph' api under the ts_backend namespace, and later we can consider a generic api at this level if we can make it work for all backends

File: torch/_lazy/__init__.py (hunk 1 above)
Comment (shunting314) [reply torch/_lazy/__init__.py:35]: And for get_tensors_backend , rename to debug.dump_ir_backend(tensors) ?

I can do that in a followup PR

File: torch/_lazy/computation.py (hunk 2 above)
Comment (shunting314) [reply torch/_lazy/computation.py:4]: I thought about the same thing. The reason I'm putting the method here is we should generalize it for XLA soon. But I'm okay to move this to ts_backend.py before that . Just let me know.

File: torch/_lazy/extract_compiled_graph.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    if not isinstance(lazy_out, (tuple, list)):
+        lazy_out = (lazy_out,)
+
+    return_value_handler = ReturnValueHandler(lazy_out)
+    if debug:
+        print("Fx code:\n", model.code)
+        print("LTC IR:", lazy.get_tensors_text(lazy_out))
+
+    graph_input_tensor_ids, graph_input_ivalues = computation.get_tensors_ts_device_data_node(lazy_out)
```
Comment (wconstab) [inline torch/_lazy/extract_compiled_graph.py:164]: should we leave this file here and leave the ts_backend specific things in it?  I think there are 2 alternatives that are not that great
 - move this file and all of this related stuff to under _lazy.ts_backend module (to be most correct, even though we want this functionality to be generic later
 - or just leave it all like it is now, but add some comments saying we intend to make this file more generic in the future and replace ts_ functions with generic ones in _lazy.computation layer

File: torch/csrc/lazy/python/init.cpp [hunk 5]
```diff
@@ -17,6 +20,12 @@
 namespace torch {
 namespace lazy {
 
+struct NoGilSection {
```
Comment (wconstab) [inline torch/csrc/lazy/python/init.cpp:23]: we don't want to land NoGilSection, can you switch to `pybind11::gil_scoped_release no_gil;` like mark_step uses?

File: torch/_lazy/extract_compiled_graph.py (hunk 4 above)
Comment (shunting314) [reply torch/_lazy/extract_compiled_graph.py:164]: Since we anyway need to make things generic to all backends, I would prefer option 2.

File: torch/csrc/lazy/ts_backend/ts_backend_impl.h [hunk 6]
```diff
@@ -1,3 +1,5 @@
+#pragma once
```
Comment (wconstab) [inline torch/csrc/lazy/ts_backend/ts_backend_impl.h:1]: woah! good catch

File: torch/_lazy/__init__.py (hunk 1 above)
Comment (wconstab) [reply torch/_lazy/__init__.py:35]: yea, sgtm.  (or, just make a parameter to debug.dump_ir(tensors, format="text"/"backend")? which is cleaner?

File: torch/_lazy/computation.py (hunk 2 above)
Comment (wconstab) [reply torch/_lazy/computation.py:4]: yea, I think ideally we want it at this level, and it is unclear if we can actually deliver that easily or if it will be a little different for xla.  but lets leave it here now and just add some comments explaining the proposal to add xla here, and that this is still 'unstable/prototype'

**Comment by shunting314** [comment]
@wconstab  this PR was not generated by ghstack thus it shows 'Merging is blocked' . I create another one with ghstack: https://github.com/pytorch/pytorch/pull/75046 . Sorry about the confusing. 

I'll resolve comments from this PR in the new one

**Comment by wconstab** [comment]
> @wconstab this PR was not generated by ghstack thus it shows 'Merging is blocked' . I create another one with ghstack: #75046 . Sorry about the confusing.
> 
> I'll resolve comments from this PR in the new one

ok, no problem.  But i'm confused, i never use ghstack but i seem to be able to land PRs that I just directly push.  did something change since this morning?

**Comment by shunting314** [comment]
@wconstab  I don't know. I just figured that since I'm merging to master, I need follow some doc which says I should use ghstack.   Maybe not using ghstack should also work.

**Comment by wconstab** [comment]
ah. you don't have to do it any particular way.  you can export from sandcastle to gh, or you can push to gh any way and import to sandcastle.  you still have to land via sandcastle either way, until GH first workflow is live.

**Comment by shunting314** [comment]
> ah. you don't have to do it any particular way. you can export from sandcastle to gh, or you can push to gh any way and import to sandcastle. you still have to land via sandcastle either way, until GH first workflow is live.

Got it. That's good to now. I guess I'm just scared by the red 'Merging is blocked' warning.
